From 80234966d717982d3e0b954fdc786f5321c66e68 Mon Sep 17 00:00:00 2001
From: Thorsten Glaser <tg@mirbsd.de>
Date: Sat, 4 Jan 2014 16:00:15 +0000
Subject: Use xsel instead of xclip

This allows writing to both PRIMARY and CLIPBOARD selections when --clip
is used, and keeping PRIMARY in memory.

Bug-Debian: http://bugs.debian.org/725653
Forwarded: no
Last-Update: 2014-01-04

Patch-Name: better-clipboard.patch
---
 README                |  4 ++--
 man/pass.1            |  6 +++---
 src/password-store.sh | 22 +++++++++++-----------
 3 files changed, 16 insertions(+), 16 deletions(-)

diff --git a/README b/README
index 47ed64b..fe27ace 100644
--- a/README
+++ b/README
@@ -19,8 +19,8 @@ Depends on:
   http://www.gnupg.org/
 - git
   http://www.git-scm.com/
-- xclip
-  http://sourceforge.net/projects/xclip/
+- xsel
+  http://www.vergenet.net/~conrad/software/xsel/
 - pwgen
   http://sourceforge.net/projects/pwgen/
 - tree
diff --git a/man/pass.1 b/man/pass.1
index 9e7e6e6..9f7e23d 100644
--- a/man/pass.1
+++ b/man/pass.1
@@ -72,7 +72,7 @@ program. This command is alternatively named \fBlist\fP.
 Decrypt and print a password named \fIpass-name\fP. If \fI--clip\fP or \fI-c\fP
 is specified, do not print the password but instead copy the first line to the
 clipboard using
-.BR xclip (1)
+.BR xsel (1)
 and then restore the clipboard after 45 seconds.
 .TP
 \fBinsert\fP [ \fI--echo\fP, \fI-e\fP | \fI--multiline\fP, \fI-m\fP ] [ \fI--force\fP, \fI-f\fP ] \fIpass-name\fP
@@ -99,7 +99,7 @@ of length \fIpass-length\fP and insert into \fIpass-name\fP. If \fI--no-symbols\
 is specified, do not use any non-alphanumeric characters in the generated password.
 If \fI--clip\fP or \fI-c\fP is specified, do not print the password but instead copy
 it to the clipboard using
-.BR xclip (1)
+.BR xsel (1)
 and then restore the clipboard after 45 seconds. Prompt before overwriting an existing password,
 unless \fI--force\fP or \fI-f\fP is specified.
 .TP
@@ -345,7 +345,7 @@ The location of the text editor used by \fBedit\fP.
 .BR gpg2 (1),
 .BR pwgen (1),
 .BR git (1),
-.BR xclip (1).
+.BR xsel (1).
 
 .SH AUTHOR
 .B pass
diff --git a/src/password-store.sh b/src/password-store.sh
index 2c5c7b4..ac7c782 100755
--- a/src/password-store.sh
+++ b/src/password-store.sh
@@ -87,17 +87,19 @@ yesno() {
 # BEGIN Platform definable
 #
 clip() {
-	# This base64 business is a disgusting hack to deal with newline inconsistancies
-	# in shell. There must be a better way to deal with this, but because I'm a dolt,
-	# we're going with this for now.
-
-	before="$(xclip -o -selection clipboard | base64)"
-	echo -n "$1" | xclip -selection clipboard
+	before_clipboard=$(xsel -o -b)
+	before_primary=$(xsel -o -p)
+	printf '%s' "$1" | xsel -i -b
+	printf '%s' "$1" | xsel -i -p
 	(
 		sleep 45
-		now="$(xclip -o -selection clipboard | base64)"
-		if [[ $now != $(echo -n "$1" | base64) ]]; then
-			before="$now"
+		now_clipboard=$(xsel -o -b)
+		now_primary=$(xsel -o -p)
+		if [[ $now_clipboard = "$1" ]]; then
+			printf '%s' "$before_clipboard" | xsel -i -b
+		fi
+		if [[ $now_primary = "$1" ]]; then
+			printf '%s' "$before_primary" | xsel -i -p
 		fi
 
 		# It might be nice to programatically check to see if klipper exists,
@@ -108,8 +110,6 @@ clip() {
 		# Clipboard managers frequently write their history out in plaintext,
 		# so we axe it here:
 		qdbus org.kde.klipper /klipper org.kde.klipper.klipper.clearClipboardHistory &>/dev/null
-
-		echo "$before" | base64 -d | xclip -selection clipboard
 	) & disown
 	echo "Copied $2 to clipboard. Will clear in 45 seconds."
 }
